File: pm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
print('---- 位相余裕と安定性の確認 -----')

import numpy as np
import control as ctrl


def cal_pm2(a, b, k):
    L = ctrl.tf([k], [1, (a + b), a * b, 0])
    gm, pm, Wcg, Wcp = ctrl.margin(L)
    print('--------------------------------------------------------------')
    print(f'位相余裕[deg]={pm:.3f}  ゲイン交差周波数={Wcp:.3f}  a={a:.3f}  b={b:.3f}  k={k:.3f}')
    try:
        Gm_db = 20 * np.log10(gm) if (np.isfinite(gm) and gm > 0) else np.inf
    except Exception:
        Gm_db = 20 * np.log10(np.abs(gm))
    print(f'ゲイン余裕[dB]={Gm_db:.3f}  位相交差周波数={Wcg:.3f}')
    FB = ctrl.feedback(L, ctrl.tf([1], [1]))
    poles = ctrl.poles(FB)
    def _fmt_pole(p):
        try:
            re = p.real
            im = p.imag
        except Exception:
            # fallback for real numbers
            return f'{p:.3f}'
        if np.isclose(im, 0):
            return f'{re:.3f}'
        return f'{re:.3f}{im:+.3f}j'

    formatted_poles = [_fmt_pole(p) for p in poles]
    print('閉ループ極 =', formatted_poles)
    return {'L': L, 'gm': gm, 'pm': pm, 'Wcg': Wcg, 'Wcp': Wcp, 'poles': poles}


def cal_pm3(a, b, c, k):
    La = ctrl.tf([1], [1, a])
    Lb = ctrl.tf([1], [1, b])
    Lc = ctrl.tf([1], [1, c])
    L = k * La * Lb * Lc
    gm, pm, Wcg, Wcp = ctrl.margin(L)
    print('--------------------------------------------------------------')
    print(f'位相余裕[deg]={pm:.3f}  ゲイン交差周波数={Wcp:.3f}  a={a:.3f}  b={b:.3f}  c={c:.3f}  k={k:.3f}')
    try:
        Gm_db = 20 * np.log10(gm) if (np.isfinite(gm) and gm > 0) else np.inf
    except Exception:
        Gm_db = 20 * np.log10(np.abs(gm))
    print(f'ゲイン余裕[dB]={Gm_db:.3f}  位相交差周波数={Wcg:.3f}')
    FB = ctrl.feedback(L, ctrl.tf([1], [1]))
    poles = ctrl.poles(FB)
    def _fmt_pole2(p):
        try:
            re = p.real
            im = p.imag
        except Exception:
            return f'{p:.3f}'
        if np.isclose(im, 0):
            return f'{re:.3f}'
        return f'{re:.3f}{im:+.3f}j'

    formatted_poles = [_fmt_pole2(p) for p in poles]
    print('閉ループ極 =', formatted_poles)
    return {'L': L, 'gm': gm, 'pm': pm, 'Wcg': Wcg, 'Wcp': Wcp, 'poles': poles}


def main():
    #　パラメータがa,b,kの場合
    examples = [
        (2, 3, 5 * np.sqrt(2)),
        (4, 6, 40 * np.sqrt(2)),
        (2, 5, 14 * np.sqrt(3)),
        (3, 3, 12 * np.sqrt(3)),
        (2 * np.sqrt(3), 2 * np.sqrt(3), 32),
        (4, 15, 114),
        (5, 9, 84),
        (np.sqrt(3), 3, np.sqrt(6) * 6),
        (1, np.sqrt(3), 2 * np.sqrt(2)),
    ]
    for a, b, k in examples:
        cal_pm2(a, b, k)

    #　パラメータがa,b,c,kの場合
    a = 1
    b = np.sqrt(3)
    c = np.sqrt(3)
    k = 4 * np.sqrt(2)
    cal_pm3(a, b, c, k)

if __name__ == '__main__':
    main()

File: pm.m
disp('---- 位相余裕と安定性の確認 -----')

function result = cal_pm2(a,b,k)
	L = tf([k],[1 (a+b) a*b 0]);
	[Gm, Pm, Wcg, Wcp]=margin(L);
	disp('--------------------------------------------------------------')
	out=['位相余裕[deg]=', num2str(Pm), '  ゲイン交差周波数=', num2str(Wcp), ...
		'  a=', num2str(a), '  b=', num2str(b), '  k=', num2str(k)];
	disp(out);
	Gm=20*log10(Gm);
	out=['ゲイン余裕[dB]=', num2str(Gm), '  位相交差周波数=', num2str(Wcg)];
	disp(out);
	H=tf([1],[1]);
   FB=feedback(L,H);
   pole(FB)  % 閉ループ極
end

function result = cal_pm3(a,b,c, k)
	La = tf([1],[1 a]);
	Lb = tf([1],[1 b]);
	Lc = tf([1],[1 c]);
	L=k*La*Lb*Lc
	[Gm, Pm, Wcg, Wcp]=margin(L);
	disp('--------------------------------------------------------------')
	out=['位相余裕[deg]=', num2str(Pm), '  ゲイン交差周波数=', num2str(Wcp), ...
		'  a=', num2str(a), '  b=', num2str(b), ' c=', num2str(c), '  k=', num2str(k)];
	disp(out);
	Gm=20*log10(Gm);
	out=['ゲイン余裕[dB]=', num2str(Gm), '  位相交差周波数=', num2str(Wcg)];
	disp(out);
	H=tf([1],[1]);
   FB=feedback(L,H);
   pole(FB)  % 閉ループ極
end

a=2; b=3; k=5*sqrt(2);
cal_pm2(a, b, k);

a=4; b=6; k=40*sqrt(2);
cal_pm2(a, b, k);

a=2; b=5; k=14*sqrt(3);
cal_pm2(a, b, k);

a=3; b=3; k=12*sqrt(3);
cal_pm2(a, b, k);

a=2*sqrt(3); b=2*sqrt(3); k=32;
cal_pm2(a, b, k);

a=4; b=15; k=114;
cal_pm2(a, b, k);

a=5; b=9; k=84;
cal_pm2(a, b, k);

a=sqrt(3); b=3; k=sqrt(6)*6;
cal_pm2(a, b, k);

a=1; b=sqrt(3); k=2*sqrt(2);
cal_pm2(a, b, k);

a=1; b=sqrt(3); c=sqrt(3); k=4*sqrt(2);
cal_pm3(a, b, c, k);
